File: python/src/problem/leetcode/medium/leetcode_24.py
# coding=utf-8

from src.data_structure.data_structure import ListNode


class Solution:
    """
    两两交换链表中的节点
    """

    def swap_pairs1(self, head: ListNode) -> ListNode:
        """
        Time: O(n), Space: O(n)

        :param head:
        :return:
        """
        if not head or not head.next:
            return head
        next = head.next
        head.next = self.swap_pairs1(next.next)
        next.next = head
        return next

    def swap_pairs2(self, head: ListNode) -> ListNode:
        """
        Time: O(n), Space: O(1)

        :param head:
        :return:
        """
        pre = dummy = ListNode(0, head)
        while pre.next and pre.next.next:
            first, second = pre.next, pre.next.next
            pre.next = second
            first.next = second.next
            second.next = first
            pre = first
        return dummy.next

File: c/medium/leetcode_24.c
//
// Created by ywh on 2020/8/14.
//
#include "../dataStructure.h"

/**
 *
 * @param head
 * @return
 */
struct ListNode *swapPairs(struct ListNode *head) {
    if (!head || !head->next) {
        return head;
    }
    struct ListNode *next = head->next;
    head->next = swapPairs(next->next);
    next->next = head;
    return next;
}

/**
 *
 * @param head
 * @return
 */
struct ListNode *swapPairs2(struct ListNode *head) {
    struct ListNode dummy, *pre = &dummy;
    dummy.next = head;
    while (pre->next && pre->next->next) {
        struct ListNode *first = pre->next, *second = pre->next->next;
        pre->next = second;
        first->next = second->next;
        second->next = first;
        pre = first;
    }
}